File: apps/mobile-legacy-rn/src/services/authService.ts
import type { AuthUser, LoginResponse } from '@servis/shared';
import { env } from '@/config/env';
import { api } from './apiClient';
import { DEMO_ACCOUNTS } from '@/mocks/demoUsers';

/** Kimlik doğrulama servis arayüzü — mock ve gerçek uygulamalar bunu paylaşır. */
export interface AuthService {
  login(email: string, password: string): Promise<LoginResponse>;
  me(): Promise<AuthUser>;
  logout(): Promise<void>;
}

/** Backend olmadan çalışan mock kimlik doğrulama. */
class MockAuthService implements AuthService {
  async login(email: string, password: string): Promise<LoginResponse> {
    // Gerçekçi bir gecikme
    await delay(400);
    const account = DEMO_ACCOUNTS.find(
      (a) => a.user.email.toLowerCase() === email.trim().toLowerCase(),
    );
    if (!account || account.password !== password) {
      throw new Error('E-posta veya şifre hatalı.');
    }
    return {
      user: account.user,
      tokens: {
        access_token: `mock-access-${account.user.id}`,
        refresh_token: `mock-refresh-${account.user.id}`,
        token_type: 'bearer',
        expires_in: 900,
      },
    };
  }

  async me(): Promise<AuthUser> {
    await delay(150);
    // Mock modda oturum kullanıcısı authStore tarafından tutulur.
    throw new Error('Mock modda /me authStore üzerinden okunur.');
  }

  async logout(): Promise<void> {
    await delay(100);
  }
}

/** Gerçek backend'e bağlanan kimlik doğrulama. */
class ApiAuthService implements AuthService {
  async login(email: string, password: string): Promise<LoginResponse> {
    const { data } = await api.post<LoginResponse>('/auth/login', { email, password });
    return data;
  }

  async me(): Promise<AuthUser> {
    const { data } = await api.get<AuthUser>('/auth/me');
    return data;
  }

  async logout(): Promise<void> {
    await api.post('/auth/logout');
  }
}

const delay = (ms: number) => new Promise<void>((r) => setTimeout(r, ms));

/** Yapılandırmaya göre aktif servis (mock varsayılan). */
export const authService: AuthService = env.useMock ? new MockAuthService() : new ApiAuthService();

File: apps/mobile/lib/data/services/auth_service.dart
import 'package:dio/dio.dart';

import '../../core/config/env.dart';
import '../mock/demo_users.dart';
import '../models/auth_models.dart';

/// Kimlik doğrulama servisi arayüzü — mock ve gerçek uygulamalar bunu paylaşır.
abstract class AuthService {
  Future<LoginResult> login(String email, String password);
  Future<AuthUser> me();
  Future<void> logout();
}

/// Backend olmadan çalışan mock kimlik doğrulama.
class MockAuthService implements AuthService {
  @override
  Future<LoginResult> login(String email, String password) async {
    await Future<void>.delayed(const Duration(milliseconds: 400));
    final normalized = email.trim().toLowerCase();
    for (final acc in demoAccounts) {
      if (acc.user.email.toLowerCase() == normalized) {
        if (acc.password != password) throw Exception('E-posta veya şifre hatalı.');
        return LoginResult(
          user: acc.user,
          tokens: AuthTokens(
            accessToken: 'mock-access-${acc.user.id}',
            refreshToken: 'mock-refresh-${acc.user.id}',
            expiresIn: 900,
          ),
        );
      }
    }
    throw Exception('E-posta veya şifre hatalı.');
  }

  @override
  Future<AuthUser> me() async => throw Exception('Mock modda /me oturum durumundan okunur.');

  @override
  Future<void> logout() async => Future<void>.delayed(const Duration(milliseconds: 100));
}

/// Gerçek backend'e bağlanan kimlik doğrulama.
class ApiAuthService implements AuthService {
  ApiAuthService(this._dio);
  final Dio _dio;

  @override
  Future<LoginResult> login(String email, String password) async {
    final res = await _dio.post<Map<String, dynamic>>('/auth/login',
        data: {'email': email, 'password': password});
    return LoginResult.fromJson(res.data!);
  }

  @override
  Future<AuthUser> me() async {
    final res = await _dio.get<Map<String, dynamic>>('/auth/me');
    return AuthUser.fromJson(res.data!);
  }

  @override
  Future<void> logout() async => _dio.post<void>('/auth/logout');
}

/// Yapılandırmaya göre aktif servis (mock varsayılan).
AuthService createAuthService(Dio dio) =>
    Env.useMock ? MockAuthService() : ApiAuthService(dio);
